src/hotspot/share/opto/inlinetypenode.cpp

```diff
@@ -460,69 +471,31 @@ void InlineTypeNode::store_flattened(GraphKit* kit, Node* base, Node* ptr, ciIns
-        kit->set_memory(store, adr_type);
-
-        if (!bundle_size_supported) {
-          kit->env()->record_method_not_compilable("Mutifield bundle size not supported for target", false);
-        }
-      } else {
-        if (!value->is_InlineType()) {
-          // Recursively store the flattened inline type field
-          value = InlineTypeNode::make_from_oop(kit, value, ft->as_inline_klass());
-        }
-        value->as_InlineType()->store_flattened(kit, base, ptr, holder, offset, decorators);
+      if (!value->is_InlineType()) {
+        // Recursively store the flattened inline type field
+        value = InlineTypeNode::make_from_oop(kit, value, ft->as_inline_klass());
       }
+      value->as_InlineType()->store_flattened(kit, base, ptr, holder, offset, decorators);
     } else {
-      // Store field value to memory
-      const TypePtr* adr_type = field_adr_type(base, offset, holder, decorators, kit->gvn());
-      Node* adr = kit->basic_plus_adr(base, ptr, offset);
+      int vec_len = ft->bundle_size();
       BasicType bt = type2field[ft->basic_type()];
-      assert(is_java_primitive(bt) || adr->bottom_type()->is_ptr_to_narrowoop() == UseCompressedOops, "inconsistent");
       const Type* val_type = Type::get_const_type(ft);
       const TypeAryPtr* ary_type = kit->gvn().type(base)->isa_aryptr();
       if (ary_type != NULL) {
         decorators |= IS_ARRAY;
       }
-      if (ft->bundle_size() > 1) {
-        int vec_len = ft->bundle_size();
-        BasicType elem_bt = ft->basic_type();
-
-        bool bundle_size_supported =
-          Matcher::match_rule_supported_vector(Op_StoreVector, vec_len, elem_bt) &&
-          Matcher::match_rule_supported_vector(VectorNode::replicate_opcode(elem_bt), vec_len, elem_bt);
-
-        // Set the vector length to  maximal supported vector length
-        // to allow graceful compilation exit at a later stage.
-        vec_len = bundle_size_supported ? vec_len : Matcher::max_vector_size(elem_bt);
-
-        // Handling for non-flattened case, with default InlineFieldMaxFlatSize of 128
-        // all the concrete vectors should be fully flattened.
-        value = value->bottom_type()->isa_vect() ? value : kit->gvn().transform(VectorNode::scalar2vector(value, vec_len, val_type, false));
-        assert(value->bottom_type()->isa_vect() && value->bottom_type()->is_vect()->length() == (uint)ft->bundle_size(), "");
-
+      const TypePtr* adr_type = field_adr_type(base, offset, holder, decorators, kit->gvn());
+      Node* adr = kit->basic_plus_adr(base, ptr, offset);
+      assert(is_java_primitive(bt) || adr->bottom_type()->is_ptr_to_narrowoop() == UseCompressedOops, "inconsistent");
+      if (value->bottom_type()->isa_vect()) {
         Node* store = kit->gvn().transform(StoreVectorNode::make(0, kit->control(), kit->memory(adr), adr, adr_type, value, vec_len));
```

Comment:
Does it need to check the Op_StoreVector supported or not here?

Reply:
Checks are done at the time of inline type IR creation. 

> Thanks for the updating! So the correspond `expand_vunbox_node` can also be implemented by calling `InlineTypeNode::make_from_oop` ?

You are correct, every time C2 see a value/primitive instance it does create an InlineTypeNode IR by loading each fields from oop using make_from_oop, thus vector inline expanders should always receive a InlineType node.

![image](https://user-images.githubusercontent.com/59989778/230518402-a0f048a4-1b1d-4379-aaaf-4d12376abd70.png)